File: app/src/main/java/org/tensorflow/demo/StylizeActivity.java
/*
 * Copyright 2017 The TensorFlow Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tensorflow.demo;

import android.Manifest;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Rect;
import android.os.Build;
import android.os.Bundle;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.View.OnTouchListener;
import android.widget.Button;
import android.widget.GridView;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.google.firebase.ml.vision.common.FirebaseVisionImage;
import com.google.firebase.ml.vision.common.FirebaseVisionImageMetadata;
import com.otaliastudios.cameraview.CameraView;
import com.otaliastudios.cameraview.frame.Frame;
import com.otaliastudios.cameraview.frame.FrameProcessor;

import timber.log.Timber;

/**
 * Sample activity that stylizes the camera preview according to "A Learned Representation For
 * Artistic Styles" (https://arxiv.org/abs/1610.07629)
 */
public class StylizeActivity extends AppCompatActivity implements FrameProcessor {

    private static final int PERMISSIONS_REQUEST = 1;

    private static final String PERMISSION_CAMERA = Manifest.permission.CAMERA;

    private CameraView cameraView;

    // Whether to actively manipulate non-selected sliders so that sum of activations always appears
    // to be 1.0. The actual style input tensor will be normalized to sum to 1.0 regardless.
    private static final boolean NORMALIZE_SLIDERS = true;

    private static final int[] SIZES = {128, 192, 256, 384, 512, 720};

    // Start at a medium size, but let the user step up through smaller sizes so they don't get
    // immediately stuck processing a large image.
    private int desiredSizeIndex = -1;
    private int desiredSize = 720;
    private int initializedSize = 0;

    private Bitmap croppedBitmap = null;

    private boolean computing = false;

    private Matrix frameToCropTransform;

    private int lastOtherStyle = 1;

    private boolean allZero = false;

    private ImageGridAdapter adapter;

    private ImageView overlay;

    private Stylize stylize;

    private Button sizeButton;

    private final OnTouchListener gridTouchAdapter = new OnTouchListener() {
        ImageSlider slider = null;

        @Override
        public boolean onTouch(final View v, final MotionEvent event) {
            switch (event.getActionMasked()) {
                case MotionEvent.ACTION_DOWN:
                    for (int i = 0; i < Styles.getCount(); ++i) {
                        final ImageSlider child = adapter.getItem(i);
                        final Rect rect = new Rect();
                        child.getHitRect(rect);
                        if (rect.contains((int) event.getX(), (int) event.getY())) {
                            slider = child;
                            slider.setHighlight(true);
                        }
                    }
                    break;

                case MotionEvent.ACTION_MOVE:
                    if (slider != null) {
                        final Rect rect = new Rect();
                        slider.getHitRect(rect);

                        final float newSliderVal = (float) Math.min(1.0, Math.max(0.0, 1.0 - (event.getY() - slider.getTop()) / slider.getHeight()));

                        setStyle(slider, newSliderVal);
                    }
                    break;

                case MotionEvent.ACTION_UP:
                    if (slider != null) {
                        slider.setHighlight(false);
                        slider = null;
                    }
                    break;

                default: // fall out

            }
            return true;
        }
    };

    @Override
    public void onCreate(final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_stylize);

        stylize = new Stylize(getAssets());

        cameraView = findViewById(R.id.cameraView);
        overlay = findViewById(R.id.preview);
        sizeButton = findViewById(R.id.sizeButton);
        sizeButton.setText("" + desiredSize);
        sizeButton.setOnClickListener(new OnClickListener() {

            @Override
            public void onClick(final View v) {
                desiredSizeIndex = (desiredSizeIndex + 1) % SIZES.length;
                desiredSize = SIZES[desiredSizeIndex];
                sizeButton.setText("" + desiredSize);
                sizeButton.postInvalidate();
            }
        });

        adapter = new ImageGridAdapter(this, Styles.getThumbnails());
        GridView grid = findViewById(R.id.grid_layout);
        grid.setAdapter(adapter);
        grid.setOnTouchListener(gridTouchAdapter);

        setStyle(adapter.getItem(0), 1.0f);

        if (hasPermission()) {
            bindCamera();
        } else {
            requestPermission();
        }
    }

    @Override
    public void process(@NonNull Frame frame) {
        if (computing) return;
        computing = true;

        int previewWidth = frame.getSize().getWidth();
        int previewHeight = frame.getSize().getHeight();

        if (desiredSize != initializedSize) {
            Timber.i("Initializing at size preview size %dx%d, stylize size %d",
                    previewWidth, previewHeight, desiredSize);
            croppedBitmap = Bitmap.createBitmap(desiredSize, desiredSize, Bitmap.Config.ARGB_8888);

            frameToCropTransform =
                    ImageUtils.getTransformationMatrix(
                            previewWidth, previewHeight,
                            desiredSize, desiredSize,
                            0, true);
            initializedSize = desiredSize;
        }

        Bitmap bitmap = FirebaseVisionImage
                .fromByteArray(frame.getData(), new FirebaseVisionImageMetadata.Builder()
                        .setFormat(FirebaseVisionImageMetadata.IMAGE_FORMAT_NV21)
                        .setWidth(frame.getSize().getWidth())
                        .setHeight(frame.getSize().getHeight())
                        .setRotation(FirebaseVisionImageMetadata.ROTATION_90)
                        .build())
                .getBitmap();

        final Canvas canvas = new Canvas(croppedBitmap);
        canvas.drawBitmap(bitmap, frameToCropTransform, null);

        final Bitmap stylizedImage = stylize.stylize(croppedBitmap, desiredSize);

        computing = false;

        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                overlay.setImageBitmap(stylizedImage);
            }
        });
    }

    private void setStyle(final ImageSlider slider, final float value) {
        slider.setValue(value);

        final int styleCount = Styles.getCount();

        if (NORMALIZE_SLIDERS) {
            // Slider vals correspond directly to the input tensor vals, and normalization is visually
            // maintained by remanipulating non-selected sliders.
            float otherSum = 0.0f;

            for (int i = 0; i < styleCount; ++i) {
                if (adapter.getItem(i) != slider) {
                    otherSum += adapter.getItem(i).getValue();
                }
            }

            if (otherSum > 0.0) {
                float highestOtherVal = 0;
                final float factor = otherSum > 0.0f ? (1.0f - value) / otherSum : 0.0f;
                for (int i = 0; i < styleCount; ++i) {
                    final ImageSlider child = adapter.getItem(i);
                    if (child == slider) {
                        continue;
                    }
                    final float newVal = child.getValue() * factor;
                    child.setValue(newVal > 0.01f ? newVal : 0.0f);

                    if (child.getValue() > highestOtherVal) {
                        lastOtherStyle = i;
                        highestOtherVal = child.getValue();
                    }
                }
            } else {
                // Everything else is 0, so just pick a suitable slider to push up when the
                // selected one goes down.
                if (adapter.getItem(lastOtherStyle) == slider) {
                    lastOtherStyle = (lastOtherStyle + 1) % styleCount;
                }
                adapter.getItem(lastOtherStyle).setValue(1.0f - value);
            }
        }

        final boolean lastAllZero = allZero;
        float sum = 0.0f;
        for (int i = 0; i < styleCount; ++i) {
            sum += adapter.getItem(i).getValue();
        }
        allZero = sum == 0.0f;

        for (ImageSlider item : adapter.getItems()) {
            item.setAllZero(allZero);
        }

        // Now update the values used for the input tensor. If nothing is set, mix in everything
        // equally. Otherwise everything is normalized to sum to 1.0.
        final float[] styleVals = new float[styleCount];
        for (int i = 0; i < styleCount; ++i) {
            styleVals[i] = allZero ? 1.0f / styleCount : adapter.getItem(i).getValue() / sum;

            if (lastAllZero != allZero) {
                adapter.getItem(i).postInvalidate();
            }
        }
        stylize.setStyleValues(styleVals);
    }

    ///////////////////////////////////////////////////////////////////////////////////////

    @Override
    public void onDestroy() {
        cameraView.clearFrameProcessors();
        super.onDestroy();
    }

    @Override
    public void onRequestPermissionsResult(
            final int requestCode, @NonNull final String[] permissions, @NonNull final int[] grantResults) {
        if (requestCode == PERMISSIONS_REQUEST) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                bindCamera();
            } else {
                requestPermission();
            }
        }
    }

    private boolean hasPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            return checkSelfPermission(PERMISSION_CAMERA) == PackageManager.PERMISSION_GRANTED;
        } else {
            return true;
        }
    }

    private void requestPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (shouldShowRequestPermissionRationale(PERMISSION_CAMERA)) {
                Toast.makeText(StylizeActivity.this, "Camera AND storage permission are required for this demo", Toast.LENGTH_LONG).show();
            }
            requestPermissions(new String[]{PERMISSION_CAMERA}, PERMISSIONS_REQUEST);
        }
    }

    private void bindCamera() {
        cameraView.setLifecycleOwner(this);
        cameraView.addFrameProcessor(this);
    }
}

File: app/src/main/java/org/tensorflow/demo/StylizeActivity.kt
package org.tensorflow.demo

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Rect
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import android.widget.Button
import android.widget.GridView
import android.widget.ImageView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.ml.vision.common.FirebaseVisionImage
import com.otaliastudios.cameraview.CameraListener
import com.otaliastudios.cameraview.CameraView
import com.otaliastudios.cameraview.PictureResult
import kotlin.math.max
import kotlin.math.min

/**
 * Sample activity that stylizes the camera preview according to "A Learned Representation For
 * Artistic Styles" (https://arxiv.org/abs/1610.07629)
 */
class StylizeActivity : AppCompatActivity() {

    private lateinit var stylize: Stylize

    private lateinit var cameraView: CameraView

    private lateinit var preview: ImageView

    private lateinit var adapter: ImageGridAdapter

    // Start at a medium size, but let the user step up through smaller sizes so they don't get
    // immediately stuck processing a large image.
    private val SIZES = intArrayOf(128, 192, 256, 384, 512, 720)
    private var desiredSize = SIZES[5]
    private var desiredSizeIndex = -1

    private var lastOtherStyle = 1

    private var allZero = false

    private var lastBitmap: Bitmap? = null

    private val gridTouchAdapter = object : OnTouchListener {
        private var slider: ImageSlider? = null

        override fun onTouch(v: View, event: MotionEvent): Boolean {
            when (event.actionMasked) {
                MotionEvent.ACTION_DOWN ->
                    for (i in 0 until Styles.count) {
                        val child = adapter.getItem(i)
                        val rect = Rect()
                        child.getHitRect(rect)
                        if (rect.contains(event.x.toInt(), event.y.toInt())) {
                            slider = child
                            slider?.setHighlight(true)
                        }
                    }

                MotionEvent.ACTION_MOVE ->
                    if (slider != null) {
                        val rect = Rect()
                        slider?.getHitRect(rect)

                        val newSliderVal = min(
                            1.0,
                            max(0.0, 1.0 - (event.y - slider!!.top) / slider!!.height)
                        ).toFloat()

                        setStyle(slider!!, newSliderVal)
                    }

                MotionEvent.ACTION_UP ->
                    if (slider != null) {
                        slider?.setHighlight(false)
                        slider = null
                    }
            }
            return true
        }
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_stylize)

        stylize = Stylize(assets)

        cameraView = findViewById(R.id.cameraView)
        preview = findViewById(R.id.preview)

        val sizeButton: Button = findViewById(R.id.sizeButton)
        sizeButton.text = "$desiredSize"
        sizeButton.setOnClickListener {
            desiredSizeIndex = (desiredSizeIndex + 1) % SIZES.size
            desiredSize = SIZES[desiredSizeIndex]
            sizeButton.text = "$desiredSize"
        }

        val captureButton: View = findViewById(R.id.captureButton)
        captureButton.setOnClickListener {
            cameraView.takePicture()
        }

        adapter = ImageGridAdapter(this, Styles.thumbnails)
        val grid: GridView = findViewById(R.id.grid_layout)
        grid.adapter = adapter
        grid.setOnTouchListener(gridTouchAdapter)

        setStyle(adapter.getItem(0), 1.0f)

        if (hasPermission()) {
            bindCamera()
        } else {
            requestPermission()
        }
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        if (requestCode == PERMISSIONS_REQUEST) {
            if (grantResults.all { it == PackageManager.PERMISSION_GRANTED }) {
                bindCamera()
            } else {
                requestPermission()
            }
        }
    }

    private fun hasPermission(): Boolean {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            checkSelfPermission(PERMISSION_CAMERA) == PackageManager.PERMISSION_GRANTED
        } else {
            true
        }
    }

    private fun requestPermission() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            if (shouldShowRequestPermissionRationale(PERMISSION_CAMERA)) {
                Toast.makeText(
                    this,
                    "Camera AND storage permission are required for this demo",
                    Toast.LENGTH_LONG
                ).show()
            }
            requestPermissions(arrayOf(PERMISSION_CAMERA), PERMISSIONS_REQUEST)
        }
    }

    private fun bindCamera() {
        cameraView.setLifecycleOwner(this)
        cameraView.addCameraListener(object : CameraListener() {

            override fun onPictureTaken(result: PictureResult) {
                super.onPictureTaken(result)
                result.toBitmap {
                    if (it != null) {
                        lastBitmap = it
                        val stylizedImage = getStylizedImageFrom(it)
                        preview.setImageBitmap(stylizedImage)
                    }
                }
            }
        })
    }

    public override fun onDestroy() {
        cameraView.clearFrameProcessors()
        super.onDestroy()
    }

    private fun getStylizedImageFrom(uri: Uri): Bitmap {
        val bitmap = FirebaseVisionImage
            .fromFilePath(this, uri)
            .bitmap

        lastBitmap = bitmap

        val previewWidth = bitmap.width
        val previewHeight = bitmap.height

        val frameToCropTransform = ImageUtils.getTransformationMatrix(
            previewWidth, previewHeight,
            desiredSize, desiredSize,
            0, true
        )
        val croppedBitmap = Bitmap.createBitmap(desiredSize, desiredSize, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(croppedBitmap)
        canvas.drawBitmap(bitmap, frameToCropTransform, null)

        return stylize.stylize(croppedBitmap, desiredSize)
    }

    private fun getStylizedImageFrom(bitmap: Bitmap): Bitmap {
        val previewWidth = bitmap.width
        val previewHeight = bitmap.height

        val frameToCropTransform = ImageUtils.getTransformationMatrix(
            previewWidth, previewHeight,
            desiredSize, desiredSize,
            0, true
        )
        val croppedBitmap = Bitmap.createBitmap(desiredSize, desiredSize, Bitmap.Config.ARGB_8888)
        val canvas = Canvas(croppedBitmap)
        canvas.drawBitmap(bitmap, frameToCropTransform, null)

        return stylize.stylize(croppedBitmap, desiredSize)
    }

    private fun setStyle(slider: ImageSlider, value: Float) {
        slider.value = value

        val styleCount = Styles.count

        if (NORMALIZE_SLIDERS) {
            // Slider vals correspond directly to the input tensor vals, and normalization is visually
            // maintained by remanipulating non-selected sliders.
            var otherSum = 0f

            for (i in 0 until styleCount) {
                if (adapter.getItem(i) != slider) {
                    otherSum += adapter.getItem(i).value
                }
            }

            if (otherSum > 0) {
                var highestOtherVal = 0f
                val factor = if (otherSum > 0f) (1f - value) / otherSum else 0f
                for (i in 0 until styleCount) {
                    val child = adapter.getItem(i)
                    if (child == slider) {
                        continue
                    }
                    val newVal = child.value * factor
                    child.value = if (newVal > 0.01f) newVal else 0.0f

                    if (child.value > highestOtherVal) {
                        lastOtherStyle = i
                        highestOtherVal = child.value
                    }
                }
            } else {
                // Everything else is 0, so just pick a suitable slider to push up when the
                // selected one goes down.
                if (adapter.getItem(lastOtherStyle) == slider) {
                    lastOtherStyle = (lastOtherStyle + 1) % styleCount
                }
                adapter.getItem(lastOtherStyle).value = 1f - value
            }
        }

        val lastAllZero = allZero
        var sum = 0f
        for (i in 0 until styleCount) {
            sum += adapter.getItem(i).value
        }
        allZero = sum == 0f

        for (item in adapter.items) {
            item.setAllZero(allZero)
        }

        // Now update the values used for the input tensor. If nothing is set, mix in everything
        // equally. Otherwise everything is normalized to sum to 1.0.
        val styleValues = FloatArray(styleCount)
        for (i in 0 until styleCount) {
            styleValues[i] = if (allZero) 1f / styleCount else adapter.getItem(i).value / sum

            if (lastAllZero != allZero) {
                adapter.getItem(i).postInvalidate()
            }
        }
        stylize.setStyleValues(styleValues)
    }

    companion object {

        private const val PERMISSIONS_REQUEST = 1

        private const val PERMISSION_CAMERA = Manifest.permission.CAMERA

        // Whether to actively manipulate non-selected sliders so that sum of activations always appears
        // to be 1.0. The actual style input tensor will be normalized to sum to 1.0 regardless.
        private const val NORMALIZE_SLIDERS = true
    }
}
